### ISSUE TITLE
MQTT: Fix wildcard topics

### ISSUE THREAD

**Issue by lsafelix75** [eclipse/smarthome#4333]
Following is the implementation of addconsumer. Problem that i am facing is the line with <<<-----. I treplaces all '+' with regex format which i think MQTT broker like mosquitto do not understand. I have also attached an output from mosquitto when a topic has '+' character in it.

> public boolean addConsumer(MqttMessageSubscriber subscriber) throws MqttException {
>         String topic = subscriber.getTopic();
>         // Prepare topic for regex pattern matching taking place in messageArrived.
>         topic = StringUtils.replace(StringUtils.replace(topic, "+", "[^/]*"), "#", ".*");  <<<------
>         synchronized (consumers) {
>             List<MqttMessageSubscriber> subscriberList = consumers.get(topic);
>             if (subscriberList == null) {
>                 subscriberList = new ArrayList<>();
>             }
>             subscriberList.add(subscriber);
>             consumers.put(topic, subscriberList);
>         }
>         if (isConnected()) {
>             try {
>                 client.subscribe(topic, qos);
>             } catch (org.eclipse.paho.client.mqttv3.MqttException e) {
>                 throw new MqttException(e);
>             }
>         }
>         return true;
>     }

**MOSQUITTO.log**
1506352001: Received SUBSCRIBE from openhab
1506352001:     cctvchannel/motion_alert/2/2/[^/]*/reply (QoS 0)
1506352001: openhab 0 cctvchannel/motion_alert/2/2/[^/]*/reply
1506352001: Sending SUBACK to openhab
1506352001: Received SUBSCRIBE from openhab
1506352001:     cctvchannel/motion_alert/2/3/[^/]*/reply (QoS 0)
1506352001: openhab 0 cctvchannel/motion_alert/2/3/[^/]*/reply
1506352001: Sending SUBACK to openhab
1506352006: Received SUBSCRIBE from openhab

**Comment by lsafelix75** [comment]
Refer to https://www.eclipse.org/forums/index.php/t/1089047/

**Comment by maggu2810** [comment]
@davidgraeff Can you comment why we replace a "+" in the topic? Isn't that a valid wildcard.

**Comment by davidgraeff** [comment]
This is a valid wildcard in MQTT topics, correct. The idea is that we convert the MQTT topic string to a regex string at this point, so that it can be matched later with incoming message topics. 

We need the original topic as well though, for registering it with the MQTT broker. It is time for a wildcard pattern test case, I think. I totally forgot about this use case.

**Comment by maggu2810** [comment]
> This is a valid wildcard in MQTT topics

@davidgraeff I only know about `+` and `*`. Didn't find `[^/]` in the spec either.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349842

**Comment by davidgraeff** [comment]
I never said that the regex part is a valid MQTT topic. Apparently I strongly need to improve my sentence structure, I guess :/

**Comment by maggu2810** [comment]
My comments has been about the bug detected by @lsafelix75 and that the replaced topic is used for subscription, too.

```Java
String topic = subscriber.getTopic();
topic = StringUtils.replace(StringUtils.replace(topic, "+", "[^/]"), "#", "."); <<<------
...
client.subscribe(topic, qos);
```

@davidgraeff Please don't edit your comments (except e.g. typos etc.). It is hard to follow discussions if comments has been changed later.

**Comment by davidgraeff** [comment]
Github should show a edit history, that would be helpful. I replaced "This" by "+" in my last comment, because that somehow confused you, I guess. But "This" was actually also correct, because it was a response to your question if "+" is a valid MQTT topic.

**Comment by maggu2810** [comment]
My comment has been about "this = `[^/]`". My mistake.
But with your changes my reply  does not make any sense for further read now.
In general it think it is very uncommon to change comments that has been already replied, to.

> Github should show a edit history

Didn't know that. Can you show me?

**Comment by davidgraeff** [comment]
I meant to say: Github should support to show an edit history ^^ Unfortunately it doesn't so far. If I were you, I would alter/delete your comment as well to match the conversation flow. It is not only me that edits. Sometimes your edit-after-first-post just sits for a while, because of bad connection or forgotten to click the publish button or whatever. I needed to remove some of my by-then superfluous comments to in some discussions.

**Comment by davidgraeff** [comment]
The issue is fixed already ^^. You can close it again, there was actually no need to rename :)
